clickhouse/testcontainers/clickhouse/__init__.py

```diff
@@ -23,38 +25,62 @@
     """
-    ClickHouse database container.
+    ClickHouse database container. This testcontainer defaults to exposing the TCP port of
+    ClickHouse. If you want to use the HTTP interface, specify port 8123 to be exposed.
 
     Example:
 
-        The example spins up a ClickHouse database and connects to it using the
-        :code:`clickhouse-driver`.
+        This example shows how to spin up ClickHouse.
+        It demonstrates how to connect to the *TCP* interface using :code:`clickhouse-driver`
+        and how to connect to the *HTTP* interface using :code:`clickhouse-connect`, the
+        official client library.
 
         .. doctest::
 
-            >>> import clickhouse_driver
             >>> from testcontainers.clickhouse import ClickHouseContainer
 
-            >>> with ClickHouseContainer("clickhouse/clickhouse-server:21.8") as clickhouse:
+            >>> # clickhouse_driver is a client lib that uses the TCP interface
+            >>> import clickhouse_driver
+            >>> # ClickHouseContainer exports the TCP port by default
+            >>> with ClickHouseContainer(image="clickhouse/clickhouse-server:21.8") as clickhouse:
             ...     client = clickhouse_driver.Client.from_url(clickhouse.get_connection_url())
             ...     client.execute("select 'working'")
             [('working',)]
+
+            >>> # clickhouse_connect is the official client lib, based on the HTTP interface
+            >>> import clickhouse_connect
+            >>> # If you want to use the HTTP interface, port 8123 needs to be exposed
+            >>> with ClickHouseContainer(port=8123) as clickhouse:
+            ...     client = clickhouse_connect.get_client(dsn=clickhouse.get_connection_url())
+            ...     client.query("select 'working'").result_rows
+            [('working',)]
     """
-    def __init__(self, image: str = "clickhouse/clickhouse-server:latest", port: int = 9000,
-                 username: Optional[str] = None, password: Optional[str] = None,
-                 dbname: Optional[str] = None, **kwargs) -> None:
+
+    def __init__(
+        self,
+        image: str = "clickhouse/clickhouse-server:latest",
+        port: int = 9000,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        dbname: Optional[str] = None,
+        **kwargs
+    ) -> None:
         raise_for_deprecated_parameter(kwargs, "user", "username")
         super().__init__(image=image, **kwargs)
-        self.username = username or os.environ.get("CLICKHOUSE_USER", "test")
-        self.password = password or os.environ.get("CLICKHOUSE_PASSWORD", "test")
-        self.dbname = dbname or os.environ.get("CLICKHOUSE_DB", "test")
+        self.username: str = username or os.environ.get("CLICKHOUSE_USER", "test")
+        self.password: str = password or os.environ.get("CLICKHOUSE_PASSWORD", "test")
+        self.dbname: str = dbname or os.environ.get("CLICKHOUSE_DB", "test")
         self.port = port
```

Comment:
Would it make sense to just expose both ports? Then the user can connect with `clickhouse_driver`, `clickhouse_connect` or both at the same time?

Reply:
Yes that would make sense. Can you advise me on code style: How would you express that as a `__init__` function signature with a default arg? The only thing I can come up with is changing the port type to list of int but I'm worried about backwards-compatibility.

Reply:
We could use something like `http_port` and `tcp_port` as for the azurite container (see [here](https://github.com/testcontainers/testcontainers-python/blob/main/azurite/testcontainers/azurite/__init__.py)). You can use `raise_for_deprecated_parameter` (see [here](https://github.com/testcontainers/testcontainers-python/blob/8a3a958626716d56e7526ad80716d57d2b13953d/opensearch/testcontainers/opensearch/__init__.py#L40) for an example). We can introduce a breaking change as the next release will be a major release.